tf-to-torch port of sequence input feature

jimthompson5802 opened the pull request:
# Code Pull Requests

Partial working port of Sequence Input Features to PyTorch.  For this initial commit, 2 of the 21 unit tests related to sequence input features are passing.  Though the "passing" tests represent bare-bone implementation.  Additional work is needed.

For the record:
```
root@14d45d5c2ea4:/opt/project/sandbox/pytest_area# pytest -v /opt/project/tests/integration_tests/test_sequence_features.py::test_sequence_encoders
================================================================= test session starts =================================================================
platform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 21 items

../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-rnn] FAILED                                               [  4%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-gru] FAILED                                               [  9%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-lstm] FAILED                                              [ 14%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-rnn] PASSED                                                 [ 19%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-gru] PASSED                                                 [ 23%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-lstm] FAILED                                                [ 28%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-rnn] FAILED                                        [ 33%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-gru] FAILED                                        [ 38%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-lstm] FAILED                                       [ 42%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-rnn] FAILED                                              [ 47%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-gru] FAILED                                              [ 52%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-lstm] FAILED                                             [ 57%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-rnn] FAILED                                [ 61%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-gru] FAILED                                [ 66%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-lstm] FAILED                               [ 71%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-rnn] FAILED                                         [ 76%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-gru] FAILED                                         [ 80%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-lstm] FAILED                                        [ 85%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-rnn] FAILED                                         [ 90%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-gru] FAILED                                         [ 95%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-lstm] FAILED                                        [100%]

```

jimthompson5802 commented:
Now have initial working version of cell_type LSTM for StackedRNN encoder.  3 out of 21 encoder tests now pass
```
platform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 21 items

../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-rnn] FAILED                               [  4%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-gru] FAILED                               [  9%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-lstm] FAILED                              [ 14%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-rnn] PASSED                                 [ 19%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-gru] PASSED                                 [ 23%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-lstm] PASSED                                [ 28%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-rnn] FAILED                        [ 33%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-gru] FAILED                        [ 38%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-lstm] FAILED                       [ 42%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-rnn] FAILED                              [ 47%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-gru] FAILED                              [ 52%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-lstm] FAILED                             [ 57%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-rnn] FAILED                [ 61%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-gru] FAILED                [ 66%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-lstm] FAILED               [ 71%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-rnn] FAILED                         [ 76%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-gru] FAILED                         [ 80%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-lstm] FAILED                        [ 85%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-rnn] FAILED                         [ 90%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-gru] FAILED                         [ 95%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-lstm] FAILED                        [100%]

```

jimthompson5802 commented:
Initial working version for `SequenceEmbedEncoder`.  6 out of 21 encoder tests passing.
```
latform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 21 items

../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-rnn] PASSED                               [  4%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-gru] PASSED                               [  9%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-lstm] PASSED                              [ 14%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-rnn] PASSED                                 [ 19%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-gru] PASSED                                 [ 23%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-lstm] PASSED                                [ 28%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-rnn] FAILED                        [ 33%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-gru] FAILED                        [ 38%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-lstm] FAILED                       [ 42%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-rnn] FAILED                              [ 47%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-gru] FAILED                              [ 52%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-lstm] FAILED                             [ 57%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-rnn] FAILED                [ 61%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-gru] FAILED                [ 66%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-lstm] FAILED               [ 71%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-rnn] FAILED                         [ 76%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-gru] FAILED                         [ 80%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-lstm] FAILED                        [ 85%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-rnn] FAILED                         [ 90%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-gru] FAILED                         [ 95%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-lstm] FAILED                        [100%]
```

jimthompson5802 commented:
Initial  working version of `StackedCNN` encoder.  9 out of 21 encoder unit tests now passing.  I'll note additional tests are needed around sequence reduction.

```
======================================================== test session starts ========================================================
platform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 21 items

../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-rnn] PASSED                             [  4%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-gru] PASSED                             [  9%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-lstm] PASSED                            [ 14%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-rnn] PASSED                               [ 19%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-gru] PASSED                               [ 23%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-lstm] PASSED                              [ 28%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-rnn] FAILED                      [ 33%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-gru] FAILED                      [ 38%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-lstm] FAILED                     [ 42%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-rnn] FAILED                            [ 47%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-gru] FAILED                            [ 52%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-lstm] FAILED                           [ 57%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-rnn] FAILED              [ 61%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-gru] FAILED              [ 66%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-lstm] FAILED             [ 71%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-rnn] PASSED                       [ 76%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-gru] PASSED                       [ 80%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-lstm] PASSED                      [ 85%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-rnn] FAILED                       [ 90%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-gru] FAILED                       [ 95%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-lstm] FAILED                      [100%]
```

jimthompson5802 commented:
Last set of commits:
* reflect adoption of the `input_shape` and `output_shape` property design
* MVP of cnnrnn encoder
* MVP of parallel_cnn encoder

With these last set of changes 15 out of 21 encoder unit test are now passing.  

```
platform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 21 items

../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-rnn] PASSED                                                  [  4%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-gru] PASSED                                                  [  9%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-lstm] PASSED                                                 [ 14%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-rnn] PASSED                                                    [ 19%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-gru] PASSED                                                    [ 23%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-lstm] PASSED                                                   [ 28%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-rnn] PASSED                                           [ 33%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-gru] PASSED                                           [ 38%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-lstm] PASSED                                          [ 42%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-rnn] PASSED                                                 [ 47%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-gru] PASSED                                                 [ 52%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-lstm] PASSED                                                [ 57%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-rnn] FAILED                                   [ 61%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-gru] FAILED                                   [ 66%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-lstm] FAILED                                  [ 71%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-rnn] PASSED                                            [ 76%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-gru] PASSED                                            [ 80%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-lstm] PASSED                                           [ 85%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-rnn] FAILED                                            [ 90%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-gru] FAILED                                            [ 95%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-lstm] FAILED                                           [100%]
```

jimthompson5802 commented:
[Prior commit](https://github.com/jimthompson5802/ludwig/commit/339c445e580a479ad16b33dbd480e0a2afff8a7f)  provides MVP for `stacked_parallel_cnn` encoder.  As of this commit 18 out of the 21 encoder unit tests now pass.
```
===================================================================== test session starts =====================================================================
platform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 21 items

../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-rnn] PASSED                                                       [  4%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-gru] PASSED                                                       [  9%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-lstm] PASSED                                                      [ 14%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-rnn] PASSED                                                         [ 19%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-gru] PASSED                                                         [ 23%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-lstm] PASSED                                                        [ 28%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-rnn] PASSED                                                [ 33%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-gru] PASSED                                                [ 38%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-lstm] PASSED                                               [ 42%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-rnn] PASSED                                                      [ 47%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-gru] PASSED                                                      [ 52%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-lstm] PASSED                                                     [ 57%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-rnn] PASSED                                        [ 61%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-gru] PASSED                                        [ 66%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-lstm] PASSED                                       [ 71%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-rnn] PASSED                                                 [ 76%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-gru] PASSED                                                 [ 80%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-lstm] PASSED                                                [ 85%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-rnn] FAILED                                                 [ 90%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-gru] FAILED                                                 [ 95%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-lstm] FAILED                                                [100%]
```

jimthompson5802 commented:
[This commit](https://github.com/jimthompson5802/ludwig/commit/386e8211d10eec884aef7568b532ecbe37da7cdd) implements the MVP for transformer encoder.  With this change all 21 encoder unit tests now pass.


results for `test_sequence_features.py::test_sequence_encoders`
```
=========================================================== test session starts ============================================================
platform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 21 items

../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-rnn] PASSED                                    [  4%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-gru] PASSED                                    [  9%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[embed-lstm] PASSED                                   [ 14%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-rnn] PASSED                                      [ 19%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-gru] PASSED                                      [ 23%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[rnn-lstm] PASSED                                     [ 28%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-rnn] PASSED                             [ 33%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-gru] PASSED                             [ 38%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[parallel_cnn-lstm] PASSED                            [ 42%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-rnn] PASSED                                   [ 47%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-gru] PASSED                                   [ 52%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[cnnrnn-lstm] PASSED                                  [ 57%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-rnn] PASSED                     [ 61%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-gru] PASSED                     [ 66%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_parallel_cnn-lstm] PASSED                    [ 71%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-rnn] PASSED                              [ 76%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-gru] PASSED                              [ 80%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[stacked_cnn-lstm] PASSED                             [ 85%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-rnn] PASSED                              [ 90%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-gru] PASSED                              [ 95%]
../../tests/integration_tests/test_sequence_features.py::test_sequence_encoders[transformer-lstm] PASSED                             [100%]
```

jimthompson5802 commented:
Summary of implemented changes.
* ported all 7 sequence encoders from TF to Torch
* added an expanded unit test (`test_sequence_encoders.py`) for the sequence encoders.  This increased the number of test cases from 21 to 504.  This expanded test cases now cover these combination of parameter settings. If this expanded unit test is accepted, I'll submit a separate PR to retire the existing one in `test_sequece_features.py::test_sequence_encoders`.
```
pytest.mark.parametrize('enc_reduce_output', [None, 'sum'])
@pytest.mark.parametrize('enc_norm', [None, 'batch', 'layer'])
@pytest.mark.parametrize('enc_num_layers', [1, 2])
@pytest.mark.parametrize('enc_dropout', [0, 0.2])
@pytest.mark.parametrize('enc_cell_type', ['rnn', 'gru', 'lstm'])
@pytest.mark.parametrize('enc_encoder', ENCODERS)
def test_sequence_encoders(
        enc_encoder,
        enc_cell_type,
        enc_dropout,
        enc_num_layers,
        enc_norm,
        enc_reduce_output
):
    # update encoder parameters for specific unit test case
    encoder_parameters['cell_type'] = enc_cell_type
    encoder_parameters['dropout'] = enc_dropout
    encoder_parameters['num_layers'] = enc_num_layers
    encoder_parameters['norm'] = enc_norm
    encoder_parameters['reduce_output'] = enc_reduce_output
```

jimthompson5802 commented:
Need confirm that  I'm using `nn.ModuleList()` in the appropriate places.  Marking PR back in DRAFT model

jimthompson5802 commented:
Completed convention to `nn.ModuleList()`.  Second try at Ready for review.

jimthompson5802 commented:
Working on out integration with fixed `FCLayer`.

jimthompson5802 commented:
Resolved issue with `stacked_parallel_cnn` encoder and `FCLayer`.   Ready for review.

w4nderlust commented on ludwig/modules/convolutional_modules.py:
```diff
@@ -392,14 +482,16 @@ def call(self, inputs, training=None, mask=None):
             )
 
-        return hidden
+        return hidden  # (batch_size, len(parallel_layers) * out_channels, seq_size)
```
this should be batch, seq_length, parallel_layers * out_channels

jimthompson5802 replied:
corrected comment on both inputs and output hidden shape

w4nderlust commented on ludwig/encoders/sequence_encoders.py:
```diff
@@ -208,7 +213,11 @@ def __init__(
-    def call(self, inputs, training=None, mask=None):
+    def get_output_shape(self, input_shape):
+        # assume input shape (seq_size, )]\
+        return input_shape[0], self.embedding_size
```
shouldn't this be a torch.Size ?

jimthompson5802 replied:
corrected when overriding the both the input_shape and output_shape address issue of "fake" tensor not being the correct dtype for sequence encoders.

w4nderlust commented on ludwig/encoders/sequence_encoders.py:
```diff
@@ -2040,6 +2113,9 @@ def __init__(
 
         logger.debug('  TransformerStack')
         self.transformer_stack = TransformerStack(
+            # todo: confirm that input_size should be hidden_size
```
that is correct, because we are either embedding to that size or projecting to that size in the previous if-else

jimthompson5802 replied:
While reviewing this comment, noticed that sequence encoders was not updated to handle `should_embed=False`.  Made changes to support this and updated unit test to test this parameter.  Expanded unit test cover 1,008 test cases.  Runs in about 10 seconds on my laptop.

jimthompson5802 commented on ludwig/features/sequence_feature.py:
```diff
@@ -170,7 +169,25 @@ def call(self, inputs, training=None, mask=None):
+        return torch.Size([self.max_sequence_length])
+
+    @property
+    def output_shape(self) -> torch.Size:
```
converted to call `output_shape`.  To make this work required overriding the default `LudwigModule.output_shape` because the default version generates the "fake" tensor of type FloatTensor, the sequence encoders expect the input tensor to be of type `torch.int*`

jimthompson5802 replied:
resolved issue with `ConcatCombiner`.

jimthompson5802 commented on ludwig/modules/attention_modules.py:
```diff
@@ -15,31 +15,35 @@
+        hidden = self.fc_layer1_activation(hidden)
+        hidden = self.fc_layer2(hidden, training=training)  # [b, s, 1]
+        attention = F.softmax(hidden, dim=1)
+        geated_inputs = torch.sum(attention * inputs, dim=1)
```
corrected typo

jimthompson5802 commented on ludwig/modules/attention_modules.py:
```diff
@@ -99,29 +123,45 @@ def call(self, inputs, training=None, mask=None):
+        # inputs [b, s, h]
+        attn_output = self.self_attention(inputs)  # [b, s, h]
+        attn_output = self.dropout1(
+            attn_output)  # [b, s, h] # todo: do we need this..., training=training)
```
removed comment re: training

jimthompson5802 commented on ludwig/modules/recurrent_modules.py:
```diff
@@ -58,57 +60,81 @@ def __init__(
+            # rnn or gru cell type
+            final_state = final_state[-1]
+
+        # todo: determine if needed for torch implementation
```
removed obsolete code

jimthompson5802 commented on ludwig/modules/convolutional_modules.py:
```diff
@@ -138,11 +185,14 @@ def __init__(
             default_pool_size=2,
             default_pool_strides=None,
-            default_pool_padding='same',
+            default_pool_padding=0,
```
changed default back to 'same'   re-ran expanded unit tests all 504 test cases pass.

jimthompson5802 commented on ludwig/modules/convolutional_modules.py:
```diff
@@ -239,17 +293,35 @@ def __init__(
                 '(length of the sequence) equal to 0. '
-                'This means that the compination of filter_size, padding, '
+                'This means that the combination of filter_size, padding, '
                 'stride, pool_size, pool_padding and pool_stride is reduces '
```
done

jimthompson5802 commented on ludwig/modules/attention_modules.py:
```diff
@@ -132,20 +172,23 @@ def __init__(
+
         for _ in range(num_layers):
             layer = TransformerBlock(
+                input_size=input_size,
```
As I understand it, I believe it is correct.  This is how I ended up with the current solution.

These two sections of the TF1 code are relevant to this discussion:
*  Part of [constructor for `StackedTransfomer`](https://github.com/ludwig-ai/ludwig/blob/b15c49e7913129fa2b93d44ec9ab19d69aeffc58/ludwig/encoders/sequence_encoders.py#L2020-L2051). Lines 2035 to 2042 sets up a layer called `self.project_to_hidden_size` and sets the flag `self.should_project = True`
* Beginning part of the [`call()` method for `StackedTransformer`](https://github.com/ludwig-ai/ludwig/blob/b15c49e7913129fa2b93d44ec9ab19d69aeffc58/ludwig/encoders/sequence_encoders.py#L2087-L2109).   As best I can tell `self.should_project` will be True when `call()` method is invoked.  This means the layer `self.project_to_hidden_size()` will be invoked to create a tensor whose last dimension is `hidden_size`.

So when `self.transformer_stack()` is called, the input tensor (aka `hidden`) has the same last dimension as the hidden layers in Transformer Stack.  In other words, "the input size for the first one" is `hidden_size`, which is the same "hidden size for the following blocks."

Did this answer the question?

w4nderlust replied:
i see your point, and your reasoning is correct wrt the `StackedTransfomer` behavior.
That said, the `TransformerStack` should likely have a parameter for inputs_size and one for hidden_size. In most cases they will be the same, in particular when called by the `StackedTransfomer`, but there could be another Module that doesn't perform the projection beforehand, so the first block in the stack of transformers will have hidden size as output and the following one will keep on using hidden size as input size. Does this make sense?

jimthompson5802 replied:
Got it....I understand for generality the first layer to be the `input_size` and the subsequent layers are `hidden_size` and not assume from the current implementation that this will always be the case.

jimthompson5802 replied:
@w4nderlust   Just pushed commit that addresses your comment.

jimthompson5802 commented:
@w4nderlust  incorporated all comments except for [this one](https://github.com/ludwig-ai/ludwig/pull/1279#discussion_r706823026).  Let me know if this answered your question.

I updated the expanded unit test for sequence encoders:
* added `should_embed` to parameter list for testing
* added function to generate a more realistic looking synthetic tensor.  Previously synthetic tensor for testing was all zeros.  Now the synthetic tensor would look more realistic with non-zero values at the beginning of the sequence with zero values at the end of the sequence.  This should set foundation to allow us to have unit test support testing of masking.   Example illustrating what I mean
```
    # generates a realistic looking synthetic sequence tensor, i.e.
    # each sequence will have non-zero tokens at the beginning with
    # trailing zero tokens, including a max length token with a single
    # zero token at the end.  Example:
    # [
    #   [3, 5, 6, 0, 0, 0],
    #   [10, 11, 12, 13, 14, 0],   # max length sequence
    #   [32, 0, 0, 0, 0, 0]        # minimum length sequence
    # ]
```

Number of parameter combinations tested now 1,004.  Test runs in about 10 seconds on my laptop.
```
@pytest.mark.parametrize('enc_should_embed', [True, False])
@pytest.mark.parametrize('enc_reduce_output', [None, 'sum'])
@pytest.mark.parametrize('enc_norm', [None, 'batch', 'layer'])
@pytest.mark.parametrize('enc_num_layers', [1, 2])
@pytest.mark.parametrize('enc_dropout', [0, 0.2])
@pytest.mark.parametrize('enc_cell_type', ['rnn', 'gru', 'lstm'])
@pytest.mark.parametrize('enc_encoder', ENCODERS)
def test_sequence_encoders(
```

jimthompson5802 commented:
prior set of commits implement:
* refactoring of sequence encoders to use `LudwigModule.input_dtype` property
* Add unit test of passthrough sequence encoder
* Fix `TypeError` int `torch_utils.sequence_length_3D()` function

Looking into error in `reduction_module.ReduceLast` class.

jimthompson5802 commented:
Last set of commits:
* Fixed errors in `ReduceLast`, `ReduceMax` and `ReduceConcat` reducers.   
* Added unit test for Passthrough Sequence Encoder that tests these reducers `None`, 'last', 'sum', 'mean', 'max' and 'concat`.  

As of this posting, all sequence encoder unit test pass (1, 014).  These run in about 10 to 12 seconds.
